build/configure/app-configure-step.js

```diff
@@ -0,0 +1,132 @@
+//       config data, so that app could easily to add/remove/reorder steps for
+//       different situations.
+const POST_APP_BUILD_STEPS = [
+  'media-resolution',
+  'post-manifest',
+  'multilocale',
+  'copy-build-stage-data',
+  'webapp-optimize',
+  'webapp-zip'
+];
+
+/**
+ * AppConfigureStep would generate all the data for build back-end, ideally it
+ * should only manage the steps relating to app itself, like app's own
+ * buildscript and POST_APP_BUILD_STEPS.
+ *
+ * @param {object} options - BUILD_CONFIG from main Makefile.
+ * @param {string} appDir - the app's source dir path.
+ * @constructor
+ */
+var AppConfigureStep = function(options, appDir) {
+  this.appDir = appDir;
+  this.webapp = utils.getWebapp(appDir, options);
+  this.name = this.webapp.sourceDirectoryName;
+  this.options = options;
+  this.buildConfig = new BuildConfig();
+  this._preSteps = [];
+};
+
+AppConfigureStep.prototype = {
+  /**
+   * Configure start from here.
+   */
+  start: function() {
+    var options = this.options;
+    this.appDirFile = utils.getFile(this.appDir);
+    this.xpcshell = utils.getEnv('XPCSHELLSDK');
+
+    // we'll execute our current modules with stringified options.
+    this.appOptions = utils.cloneJSON(options);
+
+    this.appOptions.APP_DIR = this.appDir,
+    this.appOptions.STAGE_APP_DIR = utils.joinPath(this.options.STAGE_DIR,
+      this.appDirFile.leafName);
+    this.appOptions.webapp = this.webapp;
+    // We may encounter error while stringify manifest , put it to command line
+    // and parse it back to json object, so we simply delete these since the
+    // following steps don't need these attributes.
+    delete this.appOptions.webapp.manifest;
+    delete this.appOptions.webapp.metaData;
+
+    utils.ensureFolderExists(utils.getFile(this.appOptions.STAGE_APP_DIR));
+
+    this.buildConfig.addConfig('APP_OPTIONS', this.appOptions);
+
+    // We have to rewrite below two ENV since they're using abspath and CURDIR.
+    this.buildConfig.addConfig('GAIA_DIR',
+      options.GAIA_DIR);
+    this.buildConfig.addConfig('XPCSHELLSDK',
+      options.XPCSHELLSDK);
+
+    // Create app's own makefile and store in STAGE_APP_DIR/app.mk.
+    this.appMkPath = utils.joinPath(this.appOptions.STAGE_APP_DIR, 'Makefile');
+    this.mainMake = new DependencyGraph(this.appMkPath);
+
+    this.genAppConfig();
+
+    this.mainMake.genBackend(this.buildConfig.getOutput('makefile'));
+  },
+
+  get makeCommand() {
+    return '@make -C ' + this.appOptions.STAGE_APP_DIR;
+  },
+
+  /**
+   * In order to run buildscript under each app, we need to use below command.
+   * XXX: we should try to fix it in xpcshell-commonjs.js.
+   *
+   * @param {string} moduleName - the module we would like to execute.
+   */
+  templateCommand: function(moduleName) {
+    return '@$(XPCSHELLSDK) -f $(GAIA_DIR)/build/xpcshell-commonjs.js ' +
+    '-e "run(\'' + moduleName + '\',$(APP_OPTIONS));"';
```

Comment:
we don't treat templateCommand as default shell for Makefile is due to we still need to run basic shell commands like cp and zip sometimes.